src/python/grpcio/grpc/aio/_call.py

```diff
@@ -406,7 +406,7 @@ async def _consume_request_iterator(
                 async for request in request_iterator:
                     try:
                         await self._write(request)
-                    except AioRpcError as rpc_error:
+                    except (AioRpcError, cygrpc.InternalError) as rpc_error:
```

Comment:
Can you share more information about why we should catch the exception here? Can you add a test case to demonstrate the fixing scenario?

Here is a unit test file that I think might be helpful:  https://github.com/grpc/grpc/blob/999e36fb31ca13ef467f3b7327a282a1bc4353e1/src/python/grpcio_tests/tests_aio/unit/call_test.py#L652

Reply:
the reason it should be caught here is becaues it is almost definately NOT an exception from the iterator - it's an exception from the write.

I can't find anything in call_test that accurately represents a server error.  I think it would be useful if we had such a test harness framework - the cloest I found was in server_test, so I can pursue that further if your not able to point me at a better example.

The reason we need to catch this here is so that we don't catch it below in the bare except.  When the server errors (i.e. server code hits an Exception like a AttributeError because I'm bad at software), the await call site gets a nice error response with details filled in (good error handling server!) - but the background task here, _async_request_poller, is still running.  As soon as it tries to _write it finds that the server has closed the connection and the cython raises ExecuteBatchError which inhrets from InternalError.

In the current code, I'd argue it is NOT appropriate to catch that error below, log it as `Client request_iterator raised exception` and calling `self.cancel()` can lead to a lot of confusion if the coro waiting on the call hadn't already recieved the error response yet (i.e. in a race, _write throwing an exception could happen before the response handler gets the error response, but since we've already called cancel things *really* get messed up).  Instead logging the generic error the same way we'd previously handled the AioRpcError seems benign and in all cases I can functionally test prevents tracebacks on shutdown (we still get the debug error message in logging; but I didn't change that behavior).

Reply:
Great finding! This is indeed an oversight. When Core failed to deliver messages, it may raise an InternalError.

As for the fix, I think we should add the following catch clause to `_StreamRequestMixin._write` instead. This method is the contact point between C extension and Python stream-unary/stream-stream call objects. 

```
        except cygrpc.InternalError:
            await self._raise_for_status()
```

Why I suggest against the suppression: On the worst case scenario, Core may have a weird bug causing an InternalError without proper status code and stuff, we should still raise the InternalError to the application.